### ISSUE TITLE
Dependencies - Transitioning to Redis for Enhanced Persistent Storage Solution

### ISSUE THREAD

**Issue by stan-dot** [bluesky/bluesky#1681]
Transitioning to Redis for Enhanced Persistent Storage Solution

Expected Behavior

The project should have a robust, secure, and efficient mechanism for persistent storage across IPython sessions. An ideal solution would be widely adopted, supported, and capable of meeting our scalability and performance needs without compromising security.

Current Behavior

Currently, we utilize a makeshift solution involving historydict for syncing RE.md to a SQLite database, and have explored zict as an alternative for disk-based dict representation. However, zict's use of pickle files for long-term data persistence poses security concerns. [source](https://github.com/bluesky/bluesky/issues/1227#issuecomment-511559612) Moreover, [zict](https://github.com/dask/zict) has relatively low community adoption, indicated by its 68 GitHub stars.

Possible Solution

Given the challenges we've faced, I wonder if we might explore the possibility of integrating Redis into our workflow. Redis is well-regarded for its performance and has a large community (64.4k stars on GitHub).It has a straightforward [Python SDK](https://redis.io/docs/connect/clients/python/).  However, I'm aware that adopting Redis would be a significant change, and there are considerations we'd need to address, such as operational complexity, dependency management, and ensuring data security.

Potential Risks:

    Operational Complexity: Integrating Redis could introduce new complexities in deployment and maintenance.
    Dependency Management: Adding Redis as a dependency may impact the project's setup and installation process for new users.
    Data Security: We must ensure that using Redis does not introduce security vulnerabilities, particularly in how data is accessed and stored.

SMART Goal:
To determine the viability of Redis for our project, I propose a pilot project with the following SMART goal: "By [specific date, within 3-6 months], conduct a evaluation of Redis by implementing a prototype that addresses the current limitations of historydict and zict, assessing operational complexity, dependency management, and data security concerns, with the aim of presenting findings to the team."

Steps to Reproduce (for bugs)

N/A
Context

Moving to a more stable and secure method would greatly benefit the project's long-term sustainability and ease of use for both new and existing users.

Your Environment

N/A

**Comment by prjemian** [comment]
Persistence across reboot of the host is vital.

**Comment by stan-dot** [comment]
that is a common use case for redis https://docs.redis.com/latest/kubernetes/recommendations/persistent-volumes/

**Comment by danielballan** [comment]
I love this framing of the process. I would very much like to see Bluesky support and  recommend a Redis-based solution.

We have also been exploring how to make Redis work robustly for this application. We have tried zict in the past (and contributed one or two PRs, I think) but more recently have experimented with a new approach.

https://github.com/NSLS2/redis-json-dict

I am not sure whether it is a good approach, but it is what we are trying at some beamlines now.

Finally, given recent news, it is worth mentioning that when we say "Redis" we probably mean [Valkey](https://github.com/valkey-io/valkey), the OSS project that long-time Redis contributors have rallied to, in light of Redis abandoning its OSS legacy.

**Comment by stan-dot** [comment]
JSON is a key-value store, why store the entire JSON inside REDIS instead of using the native redis key-value mappings?

It feels like extra work and complexity

**Comment by danielballan** [comment]
Storing the entire metadata dict as one JSON blob in a single Redis key is a valid way to do this.

We have use cases for accessing or updating certain select keys, such as the currently active user proposal, and we were concerned with the performance overhead of deserializing the full JSON every time we need to access or update a single key. (Some beamlines have a lot of metadata.)

It would be useful benchmark this against a dead-simple single JSON key to make an informed decision about whether it is worth the extra complexity.

**Comment by danielballan** [comment]
Agreed on Redis/Valkey. For now, I think this is just something to be aware of and watch.

**Comment by stan-dot** [comment]
alternative library: https://github.com/grantjenks/python-diskcache

@hyperrealist what do you think?
